### PR TITLE
Update the MQTT keep-alive demo to use a mocked timer-query function

### PR DESCRIPTION
**Pull request by yourslab**
Description
-----------

Because we would like to show an example of how the coreMQTT library is used without an actual timer query function, the keep-alive demo is updated to used a mocked function that always returns 0. As such, `MQTT_ReceiveLoop` must be passed zero iterations. However, the loop terminates when PUBACK, SUBACK, or UNSUBACK is received in the event callback.

Test Steps
-----------

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c [hunk 1]
```diff
@@ -752,20 +777,23 @@ static void prvMQTTPublishToTopic( MQTTContext_t * pxMQTTContext )
...
     xMQTTPublishInfo.topicNameLength = ( uint16_t ) strlen( mqttexampleTOPIC );
     xMQTTPublishInfo.pPayload = mqttexampleMESSAGE;
     xMQTTPublishInfo.payloadLength = strlen( mqttexampleMESSAGE );
 
-    /* Send a PUBLISH packet. Packet ID is not used for a QoS0 publish. */
-    xResult = MQTT_Publish( pxMQTTContext, &xMQTTPublishInfo, 0U );
+    /* Get a unique packet id. */
+    usPublishPacketIdentifier = MQTT_GetPacketId( pxMQTTContext );
+
+    /* Send a PUBLISH packet. Packet ID is not used for a QoS1 publish. */
```
Comment (muneebahmed10) [inline FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c:791]:
```suggestion
    /* Send a PUBLISH packet. */
```

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c [hunk 2]
```diff
@@ -367,19 +355,24 @@ static TimerHandle_t xKeepAliveTimer;
 static StaticTimer_t xKeepAliveTimerBuffer;
 
 /**
- * @brief Set to true when PINGREQ is sent then false when PINGRESP is received.
+ * @brief A flag indicating whether a PUBACK from the broker was received.
  */
-static volatile bool xWaitingForPingResp = false;
```
Comment (muneebahmed10) [inline FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c:372]: Do we document the fact that we're no longer enforcing a time limit on the broker to send a PINGRESP, or that we're not enforcing it to send it at all?

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c [hunk 3]
```diff
@@ -144,7 +139,7 @@
  * absence of sending any other control packets, the client MUST send a
  * PINGREQ Packet.
  */
-#define mqttexampleKEEP_ALIVE_TIMEOUT_SECONDS        ( 60U )
+#define mqttexampleKEEP_ALIVE_TIMEOUT_SECONDS        ( 15U )
```
Comment (muneebahmed10) [inline FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c:142]: So before we reported a keep alive time of 60 seconds and ran the timer every 15 seconds, and now we report a time of 15 seconds and run the timer every 3 seconds? Is this to speed up the demo?

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c [hunk 4]
```diff
@@ -156,11 +151,17 @@
 #define mqttexampleKEEP_ALIVE_DELAY                  ( pdMS_TO_TICKS( ( ( mqttexampleKEEP_ALIVE_TIMEOUT_SECONDS / 4 ) * 1000 ) ) )
 
 /**
- * @brief Delay between MQTT publishes. Note that the receive loop also has a
- * timeout, so the total time between publishes is the sum of the two delays. The
- * keep-alive delay is added here so the keep-alive timer callback executes.
+ * @brief Time to wait before calling #MQTT_ReceiveLoop.
+ *
+ * @note This delay is deliberately chosen so that the keep-alive timer callback
+ * is invoked if an expected control packet is not received within 2 iterations.
```
Comment (muneebahmed10) [inline FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c:157]: The loop iteration delay is set to 1000 ms here, and the timer runs every `mqttexampleKEEP_ALIVE_DELAY` (3000 ms), right? Am I missing something? (2 * 1000) ≯ 3000.

**Review by abhidixi11 — DISMISSED** [review]

**Review by muneebahmed10 — CHANGES_REQUESTED** [review]

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c (hunk 4 above)
Comment (yourslab) [reply FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c:157]: How'd you get 1000? Loop iteration delay here is 15/8 = 1.875 s. Timer runs every 15/4 = 3.75 s.

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c (hunk 2 above)
Comment (yourslab) [reply FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c:372]: I just had an idea to have a separate timer for checking the status of the PINGRESP. Give me a second to implement.

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c (hunk 1 above)
Comment (yourslab) [reply FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c:791]: Thanks

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c (hunk 3 above)
Comment (yourslab) [reply FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c:142]: Right, just to speed-up the demo as everything is a function of this macro. The timer ran for 1/4th of this timeout though.

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c (hunk 4 above)
Comment (muneebahmed10) [reply FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c:157]: 15/8 = 1 and 15/4 = 3 in integer division

Comment (yourslab) [reply FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c:157]: Ahh good catch

Comment (muneebahmed10) [reply FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c:157]: That works, though you could also guard against the possibility of someone changing it to another value by always rounding up (e.g. (15+7)/8 and (15+3)/4))

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c [hunk 5]
```diff
@@ -246,11 +247,12 @@ static void prvMQTTPublishToTopic( MQTTContext_t * pxMQTTContext );
 static void prvMQTTUnsubscribeFromTopic( MQTTContext_t * pxMQTTContext );
 
 /**
- * @brief The timer query function provided to the MQTT context.
+ * @brief The timer query function provided to the MQTT context. This provides
+ * an example of how to use the library without implementing an actual timer.
```
Comment (aggarw13) [inline FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c:251]:
```suggestion
 * @brief A timer query function that always returns 0. This provides
 * an example of how to use the MQTT library without implementing an actual timer.
```

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c (hunk 4 above)
Comment (yourslab) [reply FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c:157]: Don't know why I didn't just do this earlier but I just multiplied the ticks by a factor.

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c [hunk 6]
```diff
@@ -917,19 +969,39 @@ static void prvMQTTProcessIncomingPublish( MQTTPublishInfo_t * pxPublishInfo )
...
+        /* Send PINGREQ to broker */
+        LogInfo( ( "Ping the MQTT broker." ) );
+        xTransportStatus = pxTransport->send( pxTransport->pNetworkContext,
+                                              ( void * ) xPingReqBuffer.pBuffer,
+                                              xPingReqBuffer.size );
+        configASSERT( ( size_t ) xTransportStatus == xPingReqBuffer.size );
+
+        xWaitingForPingResp = true;
+        /* Start the timer to expect a PINGRESP. */
+        xTimerStatus = xTimerStart( xPingRespTimer, 0 );
```
Comment (muneebahmed10) [inline FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c:993]: Is it fine to be calling start on this timer repeatedly without calling stop on it?

**Review by muneebahmed10 — APPROVED** [review]

File: FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c (hunk 6 above)
Comment (yourslab) [reply FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Keep_Alive/DemoTasks/KeepAliveMQTTExample.c:993]: The third parameter of `xTimerCreateStatic` allows you to specify whether the time triggers periodically or just once. For the PINGRESP, once it's trigged, it won't start again unless you do so again.

**Review by n9wxu — APPROVED** [review]

**Pull request merged**